File: handlers/inline.ts
import { Composer, InputFile } from "https://deno.land/x/grammy/mod.ts";
import { getHistory } from "../api.ts";
import { getAccessToken } from "../db.ts";
import { generate } from "../image.ts";
import { getReplyMarkup } from "../utils.ts";
import env from "../env.ts";

const composer = new Composer();

composer.on("inline_query", async (ctx) => {
  const access = await getAccessToken(ctx.from.id);

  if (!access) {
    await ctx.answerInlineQuery([], {
      cache_time: 0,
      is_personal: true,
      switch_pm_text: "Click here!",
      switch_pm_parameter: "connect",
    });
    return;
  }

  let indent = Number(ctx.inlineQuery.query.split(/\s/g)[0]);
  const history = await getHistory(access);

  if (!(indent - 1 in history)) indent = 0;
  else indent -= 1;

  const track = history[indent];

  const photo = (
    await ctx.api.sendPhoto(
      env.CACHE_CHAT_ID,
      new InputFile(
        await generate(
          track.album.cover_big,
          ctx.from.first_name,
          track.title,
          track.artist.name,
          track.album.title
        )
      )
    )
  ).photo;

  await ctx.answerInlineQuery(
    [
      {
        type: "photo",
        id: String(Date.now()),
        photo_file_id: photo[photo.length - 1].file_id,
        title: track.title,
        description: track.artist.name,
        reply_markup: getReplyMarkup(track),
      },
    ],
    { cache_time: 30, is_personal: true }
  );
});

export default composer;

File: handlers/inline.go
package handlers

import (
	"dh/api"
	"dh/env"
	"dh/helpers"
	"dh/photo"
	"errors"
	"strconv"
	"strings"

	"github.com/PaulSonOfLars/gotgbot/v2"
	"github.com/PaulSonOfLars/gotgbot/v2/ext"
	"github.com/google/uuid"
)

func inlineHandler(b *gotgbot.Bot, ctx *ext.Context) error {
	tracks, err := api.FetchHistory(ctx.EffectiveUser.Id)
	if err != nil {
		if errors.Is(err, api.ErrNotAuthorized) {
			_, err = ctx.InlineQuery.Answer(
				b,
				[]gotgbot.InlineQueryResult{},
				&gotgbot.AnswerInlineQueryOpts{CacheTime: 0, IsPersonal: true, SwitchPmText: "Authorize", SwitchPmParameter: "connect"},
			)
			return err
		}
		return err
	}
	if len(tracks) == 0 {
		_, err = ctx.InlineQuery.Answer(b, []gotgbot.InlineQueryResult{gotgbot.InlineQueryResultArticle{
			Id:    uuid.NewString(),
			Title: "Your history is empty!",
			InputMessageContent: gotgbot.InputTextMessageContent{
				MessageText: "My history is empty!",
			},
		}}, nil)
		return err
	}
	indent := 0
	args := strings.Fields(ctx.InlineQuery.Query)
	if len(args) > 0 {
		indent, _ = strconv.Atoi(args[0])
	}
	indent = helpers.EnsureIndent(indent, len(tracks))
	track := tracks[indent]
	photo, err := photo.Generate(track.Artist.Name, track.Album.CoverBig, track.Artist.Name, track.Title, ctx.EffectiveUser.FirstName)
	if err != nil {
		return err
	}
	msg, err := b.SendPhoto(env.LogChatID, photo, nil)
	if err != nil {
		return err
	}
	_, err = ctx.InlineQuery.Answer(
		b,
		[]gotgbot.InlineQueryResult{
			gotgbot.InlineQueryResultCachedPhoto{
				Id:          uuid.NewString(),
				PhotoFileId: msg.Photo[len(msg.Photo)-1].FileId,
				Title:       track.Title,
				Description: track.Artist.Name,
				ReplyMarkup: helpers.GetReplyMarkup(track),
			},
		},
		&gotgbot.AnswerInlineQueryOpts{
			CacheTime:  30,
			IsPersonal: true,
		},
	)
	return err
}
